Flaky-test: AdminApi2Test.testGetListInBundle

Flaky-test: org.apache.pulsar.broker.admin.AdminApi2Test.testGetListInBundle
Number of failures: 13

org.apache.pulsar.broker.admin.AdminApi2Test is flaky. The testGetListInBundle test method fails sporadically.

```
java.lang.AssertionError: expected [3] but found [2]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.broker.admin.AdminApi2Test.testGetListInBundle(AdminApi2Test.java:2179)
```

Usage tip: To enable automatic navigation to failure message, open the following links with CTRL/CMD-click.  
[example failure 2022-02-02T02:36:26.2084173Z](https://github.com/apache/pulsar/runs/5031138202?check_suite_focus=true?check_suite_focus=true#step:11:464)  
[example failure 2022-01-31T15:37:20.8884880Z](https://github.com/apache/pulsar/runs/5008094243?check_suite_focus=true?check_suite_focus=true#step:11:464)  
[example failure 2022-01-29T18:19:53.4080491Z](https://github.com/apache/pulsar/runs/4992450850?check_suite_focus=true?check_suite_focus=true#step:11:3643)  
[example failure 2022-01-29T18:08:32.6016091Z](https://github.com/apache/pulsar/runs/4992450850?check_suite_focus=true?check_suite_focus=true#step:11:464)  
[example failure 2022-01-29T04:56:43.5270429Z](https://github.com/apache/pulsar/runs/4989344441?check_suite_focus=true?check_suite_focus=true#step:11:464)  
[example failure 2022-01-29T04:28:12.7791629Z](https://github.com/apache/pulsar/runs/4989247349?check_suite_focus=true?check_suite_focus=true#step:11:464)  
[example failure 2022-01-29T01:58:05.2523859Z](https://github.com/apache/pulsar/runs/4988542036?check_suite_focus=true?check_suite_focus=true#step:11:464)  
[example failure 2022-01-27T12:07:50.6715032Z](https://github.com/apache/pulsar/runs/4965511753?check_suite_focus=true?check_suite_focus=true#step:11:1319)  
[example failure 2022-01-27T12:02:29.1672212Z](https://github.com/apache/pulsar/runs/4965511753?check_suite_focus=true?check_suite_focus=true#step:11:458)  
[example failure 2022-01-27T09:44:23.1201561Z](https://github.com/apache/pulsar/runs/4964006047?check_suite_focus=true?check_suite_focus=true#step:11:458)  


<details>
<summary>Full exception stacktrace</summary>
<code><pre>
java.lang.AssertionError: expected [3] but found [2]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.broker.admin.AdminApi2Test.testGetListInBundle(AdminApi2Test.java:2179)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)

</pre></code>
</details>

```
java.lang.AssertionError: expected [3] but found [1]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.broker.admin.AdminApi2Test.testGetListInBundle(AdminApi2Test.java:2179)
```

Usage tip: To enable automatic navigation to failure message, open the following links with CTRL/CMD-click.  
[example failure 2022-01-29T18:14:18.8986389Z](https://github.com/apache/pulsar/runs/4992450850?check_suite_focus=true?check_suite_focus=true#step:11:2776)  
[example failure 2022-01-29T04:38:21.7070587Z](https://github.com/apache/pulsar/runs/4989272908?check_suite_focus=true?check_suite_focus=true#step:11:464)  
[example failure 2022-01-27T09:49:54.4860267Z](https://github.com/apache/pulsar/runs/4964006047?check_suite_focus=true?check_suite_focus=true#step:11:1319)  


<details>
<summary>Full exception stacktrace</summary>
<code><pre>
java.lang.AssertionError: expected [3] but found [1]
	at org.testng.Assert.fail(Assert.java:99)
	at org.testng.Assert.failNotEquals(Assert.java:1037)
	at org.testng.Assert.assertEqualsImpl(Assert.java:140)
	at org.testng.Assert.assertEquals(Assert.java:122)
	at org.testng.Assert.assertEquals(Assert.java:907)
	at org.testng.Assert.assertEquals(Assert.java:917)
	at org.apache.pulsar.broker.admin.AdminApi2Test.testGetListInBundle(AdminApi2Test.java:2179)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:132)
	at org.testng.internal.InvokeMethodRunnable.runOne(InvokeMethodRunnable.java:45)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:73)
	at org.testng.internal.InvokeMethodRunnable.call(InvokeMethodRunnable.java:11)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)

</pre></code>
</details>

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java
@@ -1099,13 +1099,24 @@ public class NamespaceService implements AutoCloseable {
         if (topic.isPersistent()) {
             return pulsar.getPulsarResources().getTopicResources().persistentTopicExists(topic);
         } else {
-            // only checks and don't do any topic creating and loading.
-            CompletableFuture<Optional<Topic>> topicFuture =
-                    pulsar.getBrokerService().getTopics().get(topic.toString());
-            if (topicFuture == null) {
-                return CompletableFuture.completedFuture(false);
+            if (topic.isPartitioned()) {
+                final TopicName partitionedTopicName = TopicName.get(topic.getPartitionedTopicName());
+                return pulsar.getBrokerService()
+                        .fetchPartitionedTopicMetadataAsync(partitionedTopicName)
+                        .thenApply((metadata) -> topic.getPartitionIndex() < metadata.partitions);
             } else {
-                return topicFuture.thenApply(Optional::isPresent).exceptionally(throwable -> false);
+                // only checks and don't do any topic creating and loading.
+                CompletableFuture<Optional<Topic>> topicFuture =
+                        pulsar.getBrokerService().getTopics().get(topic.toString());
+                if (topicFuture == null) {
+                    return CompletableFuture.completedFuture(false);
+                } else {
+                    return topicFuture.thenApply(Optional::isPresent).exceptionally(throwable -> {
+                        LOG.warn("[{}] topicFuture completed with exception when checkTopicExists, {}",
+                                topic, throwable.getMessage());
+                        return false;
+                    });
+                }
             }
         }
     }

--- a/pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java
+++ b/pulsar-broker/src/main/java/org/apache/pulsar/broker/namespace/NamespaceService.java
@@ -54,6 +54,7 @@ import org.apache.pulsar.broker.loadbalance.LoadManager;
 import org.apache.pulsar.broker.loadbalance.ResourceUnit;
 import org.apache.pulsar.broker.lookup.LookupResult;
 import org.apache.pulsar.broker.service.BrokerServiceException.ServiceUnitNotReadyException;
+import org.apache.pulsar.broker.service.Topic;
 import org.apache.pulsar.broker.service.nonpersistent.NonPersistentTopic;
 import org.apache.pulsar.broker.stats.prometheus.metrics.Summary;
 import org.apache.pulsar.broker.web.PulsarWebResource;
@@ -1098,9 +1099,14 @@ public class NamespaceService implements AutoCloseable {
         if (topic.isPersistent()) {
             return pulsar.getPulsarResources().getTopicResources().persistentTopicExists(topic);
         } else {
-            return pulsar.getBrokerService()
-                    .getTopicIfExists(topic.toString())
-                    .thenApply(optTopic -> optTopic.isPresent());
+            // only checks and don't do any topic creating and loading.
+            CompletableFuture<Optional<Topic>> topicFuture =
+                    pulsar.getBrokerService().getTopics().get(topic.toString());
+            if (topicFuture == null) {
+                return CompletableFuture.completedFuture(false);
+            } else {
+                return topicFuture.thenApply(Optional::isPresent).exceptionally(throwable -> false);
+            }
         }
     }
 

